### PR TITLE
[IOTDB-58] Replace list by array and refactor TsFileResource

### PR DESCRIPTION
**Pull request by jixuan1989**
In this PR, changes are:
BufferWriteProcessor:
- remove `fileName` (because `insertFilePath` is enough)

TsFileResource:
- remove `relativePath`
- using `autoRead` for automatically initialize `startTimeMap` and `endTimeMap`

IMemTable, AbstractMemTable:
- add returned value of `delete()`

TsRecord:
- change `dataPoints` from `List` to Array

And in many other classes, if  `Array` is enough, I replace `List<>` by `Array`.

For insertion, add a new method write(long time, Object value) to replace write(long time, String value), because the latter has to call a switch-case and something like Integer.valueOf(value).

To code reviewers:
TsFileResource is a little hard to read.

### CODE REVIEW

File: iotdb/src/main/java/org/apache/iotdb/db/engine/filenode/TsFileResource.java [hunk 1]
```diff
@@ -38,71 +47,90 @@
...
-
-    this.startTimeMap = startTimeMap;
-    this.endTimeMap = endTimeMap;
-    this.modFile = new ModificationFile(
-        Directories.getInstance().getTsFileFolder(baseDirIndex) + File.separator
-            + relativePath + ModificationFile.FILE_SUFFIX);
+  /**
+   * @param autoRead whether read the file to initialize startTimeMap and endTimeMap
+   */
+  public TsFileResource(File file, boolean autoRead) throws IOException {
```
Comment (mdf369) [inline iotdb/src/main/java/org/apache/iotdb/db/engine/filenode/TsFileResource.java]: BaseDirIndex is not init is this constructor? Won't it be used later?

Comment (jixuan1989) [reply iotdb/src/main/java/org/apache/iotdb/db/engine/filenode/TsFileResource.java]: I init it in `public TsFileResource(Map<String, Long> startTimeMap, Map<String, Long> endTimeMap,
      OverflowChangeType type, File file)` which is called by this constructor.

File: iotdb/src/main/java/org/apache/iotdb/db/engine/filenode/FileNodeProcessor.java [hunk 2]
```diff
@@ -1706,7 +1726,7 @@ public FileNodeFlushFuture flush() throws IOException {
     if (bufferWriteProcessor != null && !bufferWriteProcessor.isClosed()) {
       bufferWriteFlushFuture = bufferWriteProcessor.flush();
     }
-    if (overflowProcessor != null && !overflowProcessor.isClosed()) {
+    if (overflowProcessor != null && !bufferWriteProcessor.isClosed()) {
```
Comment (fanhualta) [inline iotdb/src/main/java/org/apache/iotdb/db/engine/filenode/FileNodeProcessor.java]: Why change it from overflowProcessor to bufferWriteProcessor?

File: iotdb/src/main/java/org/apache/iotdb/db/engine/filenode/TsFileResource.java [hunk 3]
```diff
@@ -212,45 +242,31 @@ public void removeTime(String deviceId) {
...
   public int getBaseDirIndex() {
     return baseDirIndex;
   }
 
-  public void setBaseDirIndex(int baseDirIndex) {
-    this.baseDirIndex = baseDirIndex;
-  }
+//  public void setBaseDirIndex(int baseDirIndex) {
+//    this.baseDirIndex = baseDirIndex;
+//  }
```
Comment (fanhualta) [inline iotdb/src/main/java/org/apache/iotdb/db/engine/filenode/TsFileResource.java]: Are these codes useful or not?

File: iotdb/src/main/java/org/apache/iotdb/db/engine/filenode/TsFileResource.java [hunk 4]
```diff
@@ -287,8 +303,14 @@ public TsFileResource backUp() {
 
     Map<String, Long> startTimeMapCopy = new HashMap<>(this.startTimeMap);
     Map<String, Long> endTimeMapCopy = new HashMap<>(this.endTimeMap);
-    return new TsFileResource(startTimeMapCopy, endTimeMapCopy, overflowChangeType,
-        baseDirIndex, relativePath);
+    return new TsFileResource(startTimeMapCopy,
+        endTimeMapCopy, overflowChangeType, file);
+    //return new TsFileResource(startTimeMapCopy, endTimeMapCopy, overflowChangeType,
+      //        baseDirIndex, relativePath);
```
Comment (fanhualta) [inline iotdb/src/main/java/org/apache/iotdb/db/engine/filenode/TsFileResource.java]: Please check unuseful codes  like this line.

File: iotdb/src/main/java/org/apache/iotdb/db/sync/receiver/SyncServiceImpl.java [hunk 5]
```diff
@@ -612,77 +614,39 @@ public void loadOldData(String filePath, List<String> overlapFiles)
...
         ReadOnlyTsFile readOnlyTsFile = tsfilesReaders.get(filePath);
-        ArrayList<Path> paths = new ArrayList<>();
+        List<Path> paths = new ArrayList<>();
         /** compare data with one timeseries in a round to get valid data **/
         for (String timeseries : timeseriesList) {
           paths.clear();
           paths.add(new Path(timeseries));
-          Map<InsertPlan, String> originDataPoint = new HashMap<>();
-          Map<InsertPlan, String> newDataPoint = new HashMap<>();
+          //Map<InsertPlan, String> originDataPoint = new HashMap<>();
```
Comment (fanhualta) [inline iotdb/src/main/java/org/apache/iotdb/db/sync/receiver/SyncServiceImpl.java]: Please remove these unuseful codes?

**Comment by fanhualta** [comment]
Approve your modification of class SyncServiceImpl .

File: iotdb/src/main/java/org/apache/iotdb/db/engine/filenode/FileNodeProcessor.java [hunk 6]
```diff
@@ -387,9 +385,9 @@ long getIntervalFileNodeStartTime(String deviceId) {
   /**
    * clear filenode.
    */
-  public void clearFileNode() {
+  public void clearFileNode() throws IOException {
```
[the remark this answers is not in the document]
Comment (jixuan1989) [reply iotdb/src/main/java/org/apache/iotdb/db/engine/filenode/FileNodeProcessor.java]: Hm.. this method is not written by me... @liukun4515  do you know what this function is for?

File: iotdb/src/main/java/org/apache/iotdb/db/engine/filenode/FileNodeProcessor.java [hunk 7]
```diff
@@ -549,6 +547,17 @@ public BufferWriteProcessor getBufferWriteProcessor(String processorName, long i
     return bufferWriteProcessor;
   }
 
+  /**
+   * get buffer write processor.
+   */
+  public BufferWriteProcessor getBufferWriteProcessor() throws FileNodeProcessorException {
+    if (bufferWriteProcessor == null) {
+      LOGGER.error("The bufferwrite processor is null when get the bufferwriteProcessor");
+      throw new FileNodeProcessorException("The bufferwrite processor is null");
```
[the remark this answers is not in the document]
Comment (jixuan1989) [reply iotdb/src/main/java/org/apache/iotdb/db/engine/filenode/FileNodeProcessor.java]: removed.

File: iotdb/src/main/java/org/apache/iotdb/db/engine/filenode/FileNodeProcessor.java (hunk 2 above)
Comment (jixuan1989) [reply iotdb/src/main/java/org/apache/iotdb/db/engine/filenode/FileNodeProcessor.java]: =__=!! it is a bug...

**Pull request merged**